### PR TITLE
Export jemalloc stats to prometheus

### PR DESCRIPTION
**Pull request by yiwu-arbug**
## What have you changed? (mandatory)
Exporting 6 jemalloc stats to prometheus. Details of these counters can be found in http://jemalloc.net/jemalloc.3.html (search for, e.g. "stats.allocated"). Rough meaning of them are:
* allocated: how much memory user code (the whole tikv in this case) requested.
* active: how much jemalloc allocated to satisfy the allocation requests (because jemalloc has size classes, e.g. user request for 3.9k memory, and the closest size class is 4k, so jemalloc end up allocate 4k).
* metadata: extra usage by jemalloc itself, e.g. to keep track of allocation size of each item
* resident: close to actual memory used by jemalloc.

These counters should be useful to monitor jemalloc memory fragmentation ratio.

## What are the type of the changes? (mandatory)
Improvement

## How has this PR been tested? (mandatory)
Run the code in IDC.
Grafana snapshot:
screen shot 2019-03-07 at 10 21 24 pm

## Does this PR affect documentation (docs) update? (mandatory)
Not sure if we document exported metrics.

## Does this PR affect tidb-ansible update? (mandatory)
No.

### CODE REVIEW

File: components/tikv_alloc/src/lib.rs [hunk 1]
```diff
@@ -130,6 +137,29 @@ mod imp {
...
+        pub metadata: usize,
+        pub resident: usize,
+        pub mapped: usize,
+        pub retained: usize,
+    }
+
+    pub fn fetch_stats() -> Result<JemallocStats, i32> {
+        let mut stats: JemallocStats = Default::default();
+        unsafe {
+            jemallocator::mallctl_fetch(STATS_ALLOCATED, &mut stats.allocated)?;
```
Comment (yiwu-arbug) [inline components/tikv_alloc/src/lib.rs]: jemalloc_ctl crate claim to be more efficient than `jemallocator::mallctl_fetch`. I plan to change to that if the PR look ok overall.
https://docs.rs/jemalloc-ctl/0.2.0/jemalloc_ctl/

**Comment by yiwu-arbug** [comment]
First time rust and tikv coder! Any detailed comments are welcome.

For the PR I also upgrade jemallocator. Jemallocator 0.1.9 doesn't come with stats. Please let me know if it is not okay.

**Comment by yiwu-arbug** [comment]
> Instead of starting a new thread, what about implementing the `prometheus::Collector` trait? Like this one https://github.com/tikv/tikv/blob/206209fbcc4c45ae50e1a66fc5df3ac39a6570f1/src/util/metrics/threads_linux.rs

will take a look.

File: components/tikv_alloc/src/lib.rs [hunk 2]
```diff
@@ -112,9 +113,10 @@ pub use self::imp::*;
 // The implementation of this crate when jemalloc is turned on
 #[cfg(all(unix, not(fuzzing), not(feature = "no-jemalloc")))]
 mod imp {
+    use jemalloc_ctl::{stats, Epoch};
```
Comment (zhangjinpeng1987) [inline components/tikv_alloc/src/lib.rs]: `Epoch` can be confused with raft epoch, how about 
`use jemalloc_ctl::{stats, Epoch as JeEpoch}` ?

Comment (yiwu-arbug) [reply components/tikv_alloc/src/lib.rs]: fixed

File: src/util/metrics/jemalloc_metrics.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+struct JemallocStatsCollector {
+    descs: Vec<Desc>,
+    metrics: Mutex<IntGaugeVec>,
+}
+
+impl JemallocStatsCollector {
+    fn new<S: Into<String>>(namespace: S) -> Result<JemallocStatsCollector> {
+        let ns = namespace.into();
+        let stats = IntGaugeVec::new(
+            Opts::new("jemalloc_stats", "Jemalloc stats").namespace(ns.clone()),
```
Comment (zhangjinpeng1987) [inline src/util/metrics/jemalloc_metrics.rs]: I think `ns` don't need to clone.

File: src/util/metrics/jemalloc_metrics.rs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+        })
+    }
+}
+
+impl Collector for JemallocStatsCollector {
+    fn desc(&self) -> Vec<&Desc> {
+        self.descs.iter().collect()
+    }
+
+    fn collect(&self) -> Vec<MetricFamily> {
```
Comment (zhangjinpeng1987) [inline src/util/metrics/jemalloc_metrics.rs]: How frequently this function will be called?

**Comment by zhangjinpeng1987** [comment]
Please sign-off.

File: src/util/metrics/jemalloc_metrics.rs (hunk 4 above)
Comment (yiwu-arbug) [reply src/util/metrics/jemalloc_metrics.rs]: Grafana is showing a data point every 30 seconds, but I don't really know where it is configured. @overvenus do you happen to know?

**Comment by siddontang** [comment]
@yiwu-arbug 

Don't forget to send a PR to add this in Grafana here https://github.com/pingcap/tidb-ansible

File: components/tikv_alloc/Cargo.toml [hunk 5]
```diff
@@ -23,7 +23,13 @@ log = "0.3"
 tempdir = "0.3"
 
 [dependencies.jemallocator]
+version = "0.1.9"
```
Comment (siddontang) [inline components/tikv_alloc/Cargo.toml]: PTAL @BusyJay

File: src/util/metrics/jemalloc_metrics.rs (hunk 4 above)
Comment (breeswish) [reply src/util/metrics/jemalloc_metrics.rs]: @zhangjinpeng1987 It is collected as the same frequency of our remaining metrics.

File: src/util/metrics/jemalloc_metrics.rs [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        let metrics = self.metrics.lock().unwrap();
```
Comment (breeswish) [inline src/util/metrics/jemalloc_metrics.rs]: Maybe we don't need this lock.

**Comment by yiwu-arbug** [comment]
@siddontang sure thing.

File: src/util/metrics/mod.rs [hunk 7]
```diff
@@ -26,6 +28,8 @@ mod threads_dummy;
 #[cfg(not(target_os = "linux"))]
 pub use self::threads_dummy::monitor_threads;
 
+pub use self::jemalloc_metrics::monitor_jemalloc_stats;
```
Comment (overvenus) [inline src/util/metrics/mod.rs]: Maybe we need `#[cfg(all(unix, not(fuzzing), not(feature = "no-jemalloc")))]`?

**Comment by yiwu-arbug** [comment]
@breeswish @overvenus addressed comments. mind take a look again?

**Pull request merged**